coding/wml-v20/src/main/java/org/n52/sos/decode/WmlObservationProcessDecoderv20.java

```diff
@@ -0,0 +1,235 @@
+import org.n52.sos.util.CodingHelper;
+import org.n52.sos.util.CollectionHelper;
+import org.n52.sos.util.XmlOptionsHelper;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Joiner;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Lists;
+
+import net.opengis.waterml.x20.ObservationProcessDocument;
+import net.opengis.waterml.x20.ObservationProcessPropertyType;
+import net.opengis.waterml.x20.ObservationProcessType;
+
+public class WmlObservationProcessDecoderv20 extends AbstractWmlDecoderv20 implements ProcedureDecoder<Object, Object> {
+
+	private static final Logger LOGGER = LoggerFactory.getLogger(WmlObservationProcessDecoderv20.class);
+
+	@SuppressWarnings("unchecked")
+	private static final Set<DecoderKey> DECODER_KEYS = CollectionHelper.union(
+			CodingHelper.decoderKeysForElements(WaterMLConstants.NS_WML_20_PROCEDURE_ENCODING,
+					ObservationProcessDocument.class, ObservationProcessPropertyType.class, ObservationProcessType.class),
+			CodingHelper.decoderKeysForElements(WaterMLConstants.NS_WML_20,
+					ObservationProcessDocument.class, ObservationProcessPropertyType.class, ObservationProcessType.class));
+
+	private static final Set<String> SUPPORTED_PROCEDURE_DESCRIPTION_FORMATS = Collections
+			.singleton(WaterMLConstants.NS_WML_20_PROCEDURE_ENCODING);
+
+	private static final Map<String, ImmutableMap<String, Set<String>>> SUPPORTED_TRANSACTIONAL_PROCEDURE_DESCRIPTION_FORMATS = ImmutableMap
+			.of(SosConstants.SOS, ImmutableMap.<String, Set<String>> builder()
+					.put(Sos2Constants.SERVICEVERSION, ImmutableSet.of(WaterMLConstants.NS_WML_20_PROCEDURE_ENCODING))
+					.build());
+
+	public WmlObservationProcessDecoderv20() {
+		LOGGER.debug("Decoder for the following keys initialized successfully: {}!",
+				Joiner.on(", ").join(DECODER_KEYS));
+	}
+
+	@Override
+	public Set<DecoderKey> getDecoderKeyTypes() {
+		return Collections.unmodifiableSet(DECODER_KEYS);
+	}
+
+	@Override
+	public Map<SupportedTypeKey, Set<String>> getSupportedTypes() {
+		return Collections.singletonMap(SupportedTypeKey.ProcedureDescriptionFormat,
+				SUPPORTED_PROCEDURE_DESCRIPTION_FORMATS);
+	}
+
+	@Override
+	public Set<String> getConformanceClasses() {
+		return Collections.emptySet();
+	}
+
+	@Override
+	public Set<String> getSupportedProcedureDescriptionFormats(String service, String version) {
+		if (SUPPORTED_TRANSACTIONAL_PROCEDURE_DESCRIPTION_FORMATS.containsKey(service)
+				&& SUPPORTED_TRANSACTIONAL_PROCEDURE_DESCRIPTION_FORMATS.get(service).containsKey(version)) {
+			return SUPPORTED_TRANSACTIONAL_PROCEDURE_DESCRIPTION_FORMATS.get(service).get(version);
+		}
+		return Collections.emptySet();
+	}
+
+	@Override
+	public Object decode(Object object) throws OwsExceptionReport, UnsupportedDecoderInputException {
```

Comment:
Why not splitting this class up into 3 classes and get rid of this `Object` types?